File: lib/pacer/pipe/is_unique_pipe.rb
module Pacer::Pipes
  class IsUniquePipe < RubyPipe
    import java.util.ArrayList
    import com.tinkerpop.pipes.sideeffect.SideEffectPipe
    import com.tinkerpop.pipes.util.iterators.ExpandableIterator
    import com.tinkerpop.pipes.filter.DuplicateFilterPipe

    include SideEffectPipe

    def initialize
      super()
      prepare_state
    end

    def processNextStart
      value = starts.next
      check_uniqueness value if @unique
      value
    end

    def reset
      super()
      prepare_state
    end

    def unique?
      @unique
    end

    def getSideEffect
      @unique
    end

    protected

    def check_uniqueness(value)
      @expando.add value
      @unique_pipe.next
    rescue EmptyPipe, java.util.NoSuchElementException
      @unique = false
    end

    def prepare_state
      @unique = true
      @expando = ExpandableIterator.new ArrayList.new.iterator
      @unique_pipe = DuplicateFilterPipe.new
      @unique_pipe.setStarts @expando
    end

  end
end

File: ext/src/main/java/com/xnlogic/pacer/pipes/IsUniquePipe.java
package com.xnlogic.pacer.pipes;

import java.util.LinkedHashSet;
import java.util.Set;

import com.tinkerpop.pipes.AbstractPipe;

public class IsUniquePipe<T> extends AbstractPipe<T, T> {
    private boolean unique;
    private Set<T> historySet;
  
    public IsUniquePipe() {
        super();
        this.prepareState();
    }

    protected T processNextStart() {
        T value = this.starts.next();
        if (this.unique)
            this.checkUniqueness(value);
        return value;
    }

    public void reset() {
        super.reset();
        this.prepareState();
    }
    
    public boolean isUnique() {
        return this.unique;
    }

    public boolean getSideEffect() {
        return this.unique;
    }
    
    protected void checkUniqueness(T value) {
        if (!this.historySet.add(value)) {
            this.unique = false;
            this.historySet = null;
        }
    }

    protected void prepareState() {
        this.historySet = new LinkedHashSet<T>();
        this.unique = true;
    }
}
